docs/Controllers.md

```diff
@@ -0,0 +1,87 @@
+
+## Registry Service Discovery	
+
+In order to locate the Registry, the Controller SHOULD support all of the following: 
+* Unicast DNS Service Discovery (DNS-SD).
+* mDNS.
+* Direct configuration with the location of a preferred Registry.
+
+The Controller SHOULD offer unicast DNS-SD as the default mechanism. 
+
+Controllers SHOULD observe and interpret all of the TXT records returned with the DNS Service Discovery responses according to the requirements for Query API Clients specified in the [API Paths section of the APIs document](APIs.md#api-paths) in this specification.
+
+## Query API
+The Controller MUST be capable of using the Registry's IS-04 Query API to discover any registered resource, including Node, Device,  Source, Flow, Sender, and Receiver,
+as described in the [APIs section](README.md#apis) of this specification.
+
+The Controller MUST use the Registry�s IS-04 Query API either via the REST API or by requesting WebSocket subscriptions.
+
+When using the Query API, basic queries SHOULD be used (and advanced query language where available) to cut down on the volume of resources returned to the Controller.	
+
+Controllers SHOULD adhere to the version downgrade requirements for Query API Clients specified in the [Requirements for Query API Clients section of the Upgrade Path document](Upgrade%20Path.md#requirements-for-query-api-clients) in this specification.
+
+## Pagination
+In large systems, API resources rely upon pagination in order to return high volumes of data.
+This can be both hard to keep track of for a client, and require a large number of requests in order to scan the entire data set.
+
+For this reason it is recommended to use the RESTful QUERY API for debug and development purposes only.
+
+If using the RESTful Query API rather than WebSockets, Pagination requirements MUST be implemented as specified in the [Pagination section of the APIs: Query Parameters document](APIs%20-%20Query%20Parameters.md#pagination) in this specification.
+	
+## WebSockets & Subscriptions	
+Where a WebSocket or other subscription based mechanism is provided for Controller usage, it is strongly recommended that Controllers make use of this and do not use the API resources directly.
+
+If a WebSocket connection fails, then an attempt to reconnect to the WebSocket SHOULD NOT be attempted. Instead, a new subscription SHOULD be created with this API or a different one if required. If all available APIs return errors, an exponential backoff algorithm SHOULD be used when retrying until a success code is returned.
+
+## Dynamic Update of Resources
+The Controller MUST be capable of using the Registry's IS-04 Query API to discover and dynamically update the state of any registered resource, including Node, Device,  Source, Flow, Sender, and Receiver.
+
+* The Controller MUST indicate available Senders to the user.
+* The Controller MUST dynamically indicate to the user when a Sender is put �offline� or put back 'online'.
```

Comment:
Do we have a definition for ‘offline’ and 'online ? Does it mean on the network ... active ... master enabled ...

Reply:
Agree it's unclear. Also "dynamically".

Reply:
I believe the intent of the statement is to indicate to a user when Senders disappear from the registry for some reason and then reappear at some later point, but in a timely way. @garethsb @andrewbonney Do you have suggestions for better/ more consistent wording than 'online' and 'offline', (and for that matter dynamically) that might be being used in other NMOS specifications?

Reply:
I checked JT-NM Tested test plan. Section 7.3 **Updating the list of resources dynamically from the IS-04
Registry** uses the phrases "putting offline/online" but explains what these mean (disconnecting from/reconnecting to the network). I quite like the description of the pass criteria:

> Changes in presence/absence of Reference Sender shall be reflected in the BCuT’s UI after a maximum of 30 seconds.

This wording makes clear that its about whether the Sender is currently present in the Registry or not, which is therefore dependent on the garbage collection interval, etc. (It also seems to me to permit that a Controller can simply stop showing the Sender at all when it's absent, which "indicate to the user" does perhaps not do so well?)

The [IS-04 Behaviour - Registration](https://specs.amwa.tv/is-04/releases/v1.3.1/docs/4.1._Behaviour_-_Registration.html#network-disconnect--reconnect) documentation also talks about registration and unregistration, disconnection and reconnection. This and other parts of the spec talk about adding and removing resources from the Registry.

* offline/online (and unavailable/available, inaccessible/accessible)
* disconnected/(re)connnected
* absent/present
 * removed/added
* unregistered/(re)registered

I like the later ones more than the earlier ones as they are more clearly about a Sender's registration status not its actual availability on the network or its stream status (which is obviously useful but I think not the point here?).

"Dynamically", I think, is shorthand for "without requiring user instigation and in a timely manner", right?

Reply:
Current statement is now:
> * The Controller MUST reflect changes in presence/absence of Senders to the user after a maximum of 30 seconds.

So I think this has been dealt with... There's a comment https://github.com/AMWA-TV/is-04/pull/179#discussion_r848087550 to discuss why this language isn't the same for both Senders and Receivers.

Reply:
Resolving conversation as now outdated.